File: packages/python/tracep/src/tracep/tracer.py
"""
tracep.tracer
=============
Fire-and-forget OTel tracer that:
  - exports via OTLP/HTTP JSON to <endpoint>/v1/traces on port 4318
  - uses BatchSpanProcessor (batch=50, flush every 1 s)
  - sends all trace() calls through a background daemon queue thread
  - retries exports 3× with 100 ms / 200 ms / 400 ms back-off
  - never blocks the caller
"""

from __future__ import annotations

import logging
import queue
import threading
from typing import Dict, Optional, Tuple

from opentelemetry import context as otel_context
from opentelemetry import trace as otel_trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import ReadableSpan, Span, TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor, SpanExportResult
from opentelemetry.sdk.trace.export.in_memory_span_exporter import InMemorySpanExporter
from opentelemetry.trace import NonRecordingSpan, SpanContext, SpanKind, TraceFlags
from opentelemetry.trace.status import Status, StatusCode

logger = logging.getLogger(__name__)

# Public helpers so callers can use keyword-style options without raw strings.
class WithParent(str):
    """Thin str sub-class so type checkers can distinguish parent tid values."""


class WithLevel(str):
    """Thin str sub-class so type checkers can distinguish level values."""


# ---------------------------------------------------------------------------
# Internal worker command types
# ---------------------------------------------------------------------------
_CMD_TRACE = "trace"
_CMD_END   = "end"
_CMD_STOP  = "stop"


class Tracer:
    """
    Lightweight OTel tracer facade.

    Parameters
    ----------
    endpoint : str
        Base URL of the OTLP collector, e.g. ``"http://localhost:4318"``.
        The exporter will POST to ``<endpoint>/v1/traces``.
    api_key  : str
        Bearer token sent in every ``Authorization`` header.
    service  : str
        Value of the ``service.name`` OTel resource attribute.
    """

    def __init__(self, endpoint: str, api_key: str, service: str) -> None:
        self._endpoint = endpoint.rstrip("/")
        self._api_key  = api_key
        self._service  = service

        # ---- OTel provider setup -----------------------------------------
        resource = Resource.create({"service.name": service})

        exporter = OTLPSpanExporter(
            endpoint=f"{self._endpoint}/v1/traces",
            headers={"Authorization": f"Bearer {api_key}"},
        )

        processor = BatchSpanProcessor(
            exporter,
            max_export_batch_size=50,
            schedule_delay_millis=1000,
            export_timeout_millis=10_000,
            max_queue_size=2048,
        )

        self._provider = TracerProvider(resource=resource)
        self._provider.add_span_processor(processor)

        self._otel_tracer = self._provider.get_tracer(__name__)

        # ---- Per-trace state storage --------------------------------------
        # _roots : tid -> root Span
        # _spans : (tid, cls, fn) -> child Span
        self._roots: Dict[str, Span] = {}
        self._spans: Dict[Tuple[str, str, str], Span] = {}
        self._lock  = threading.Lock()

        # ---- Background worker -------------------------------------------
        self._q: "queue.Queue[dict]" = queue.Queue()
        self._worker = threading.Thread(
            target=self._run_worker, name="tracep-worker", daemon=True
        )
        self._worker.start()

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def start(self, name: str) -> str:
        """
        Start a new root trace.

        Returns
        -------
        str
            Hex-encoded 128-bit trace ID (32 hex chars), usable as *tid*
            in subsequent :meth:`trace` / :meth:`end` calls.
        """
        root_span: Span = self._otel_tracer.start_span(
            name, kind=SpanKind.INTERNAL
        )
        tid = format(root_span.get_span_context().trace_id, "032x")

        with self._lock:
            self._roots[tid] = root_span

        return tid

    def trace(
        self,
        tid: str,
        cls: str,
        function: str,
        step: str,
        message: str,
        *,
        parent: Optional[str] = None,
        level: str = "info",
    ) -> None:
        """
        Enqueue a tracing event.  Returns immediately; work is done on
        a background daemon thread.

        Parameters
        ----------
        tid      : Trace ID returned by :meth:`start`.
        cls      : Logical class / module name → ``code.namespace`` attribute.
        function : Function / method name → ``code.function`` attribute.
        step     : Name of the OTel span event.
        message  : Human-readable description, stored as event attribute.
        parent   : Optional parent span key ``"<cls>/<function>"`` form **or**
                   the string tid of another trace to link to.
        level    : Severity label (``"info"``, ``"warn"``, ``"error"`` …).
        """
        self._q.put_nowait(
            {
                "cmd":      _CMD_TRACE,
                "tid":      tid,
                "cls":      cls,
                "function": function,
                "step":     step,
                "message":  message,
                "parent":   parent,
                "level":    level,
            }
        )

    def end(self, tid: str, status: str) -> None:
        """
        Enqueue a trace-end command.  Returns immediately.

        Parameters
        ----------
        tid    : Trace ID to close.
        status : ``"ok"`` or ``"error"``.
        """
        self._q.put_nowait({"cmd": _CMD_END, "tid": tid, "status": status})

    def close(self) -> None:
        """Gracefully stop the background worker and shut down the provider."""
        self._q.put_nowait({"cmd": _CMD_STOP})
        self._worker.join(timeout=5)

    # ------------------------------------------------------------------
    # Background worker
    # ------------------------------------------------------------------

    def _run_worker(self) -> None:
        while True:
            try:
                cmd = self._q.get()
                kind = cmd.get("cmd")

                if kind == _CMD_TRACE:
                    self._do_trace(cmd)
                elif kind == _CMD_END:
                    self._do_end(cmd)
                elif kind == _CMD_STOP:
                    break
                else:
                    logger.warning("tracep: unknown command %r", kind)
            except Exception:  # pylint: disable=broad-except
                logger.exception("tracep: worker error")
            finally:
                try:
                    self._q.task_done()
                except ValueError:
                    pass  # task_done called on empty queue guard

    # ------------------------------------------------------------------
    # Actual (sync) implementations called from worker thread
    # ------------------------------------------------------------------

    def _do_trace(self, cmd: dict) -> None:
        tid      = cmd["tid"]
        cls      = cmd["cls"]
        function = cmd["function"]
        step     = cmd["step"]
        message  = cmd["message"]
        parent   = cmd.get("parent")
        level    = cmd.get("level", "info")

        with self._lock:
            root = self._roots.get(tid)

        if root is None:
            # Unknown tid — silently ignore as specified
            logger.debug("tracep: trace() called with unknown tid=%r; ignoring", tid)
            return

        key = (tid, cls, function)

        with self._lock:
            span = self._spans.get(key)

        if span is None:
            # Determine parent context
            if parent:
                parent_key = (tid, *parent.split("/", 1)) if "/" in str(parent) else None
                with self._lock:
                    parent_span = (
                        self._spans.get(tuple(parent_key))  # type: ignore[arg-type]
                        if parent_key
                        else self._roots.get(parent)
                    )
                if parent_span is not None:
                    ctx = otel_trace.set_span_in_context(parent_span)
                else:
                    ctx = otel_trace.set_span_in_context(root)
            else:
                ctx = otel_trace.set_span_in_context(root)

            span = self._otel_tracer.start_span(
                f"{cls}.{function}",
                context=ctx,
                kind=SpanKind.INTERNAL,
                attributes={
                    "code.namespace": cls,
                    "code.function":  function,
                },
            )

            with self._lock:
                self._spans[key] = span

        # Add the event regardless of whether span was just created or cached
        span.add_event(
            step,
            attributes={"message": message, "level": level},
        )

    def _do_end(self, cmd: dict) -> None:
        tid    = cmd["tid"]
        status = cmd.get("status", "ok")

        otel_status = (
            Status(StatusCode.OK)
            if status == "ok"
            else Status(StatusCode.ERROR, description=status)
        )

        with self._lock:
            child_keys = [k for k in self._spans if k[0] == tid]
            child_spans = [self._spans.pop(k) for k in child_keys]
            root = self._roots.pop(tid, None)

        # End children first (LIFO order)
        for span in reversed(child_spans):
            try:
                span.set_status(otel_status)
                span.end()
            except Exception:  # pylint: disable=broad-except
                logger.exception("tracep: error ending child span")

        if root is not None:
            try:
                root.set_status(otel_status)
                root.end()
            except Exception:  # pylint: disable=broad-except
                logger.exception("tracep: error ending root span")

        # Force flush so BatchSpanProcessor ships everything immediately
        try:
            self._provider.force_flush(timeout_millis=5000)
        except Exception:  # pylint: disable=broad-except
            logger.exception("tracep: force_flush failed")

File: packages/swift/tracep/Sources/Tracep/Tracer.swift
import Foundation
import OpenTelemetryApi
import OpenTelemetrySdk
import OpenTelemetryProtocolExporterHTTP

// ---------------------------------------------------------------------------
// MARK: - Tracer
// ---------------------------------------------------------------------------

/// A fire-and-forget OTel tracing client.
///
/// ```swift
/// let tracer = Tracer(endpoint: "http://localhost:4318",
///                     apiKey: "my-key",
///                     service: "my-service")
///
/// let tid = tracer.start(name: "user-login")
/// tracer.trace(tid: tid, cls: "AuthService", function: "login",
///              step: "validate", message: "Checking credentials")
/// tracer.end(tid: tid, status: "ok")
/// tracer.close()
/// ```
public final class Tracer {

    // MARK: - Private state

    private let tracerProvider: TracerProviderSdk
    private let otelTracer: OpenTelemetryApi.Tracer

    /// Thread-safe dictionary of active TraceContexts keyed by hex trace-id.
    private var activeTraces: [String: TraceContext] = [:]
    private let traceLock = NSLock()

    /// Background queue for all OTel operations (fire-and-forget).
    private let queue = DispatchQueue(label: "com.tracep.background", qos: .background)

    // MARK: - Init

    /// Creates a new `Tracer` and bootstraps the OTel pipeline.
    ///
    /// - Parameters:
    ///   - endpoint: Base URL of the OTLP collector, e.g. `"http://collector:4318"`.
    ///   - apiKey:   API key sent as `Authorization: Bearer <apiKey>`.
    ///   - service:  Value for the `service.name` OTel resource attribute.
    public init(endpoint: String, apiKey: String, service: String) {
        // ----------------------------------------------------------------
        // 1. Build the OTLP/HTTP exporter
        // ----------------------------------------------------------------
        let tracesEndpoint = endpoint.hasSuffix("/")
            ? "\(endpoint)v1/traces"
            : "\(endpoint)/v1/traces"

        guard let url = URL(string: tracesEndpoint) else {
            fatalError("[Tracep] Invalid endpoint URL: \(tracesEndpoint)")
        }

        let otlpConfig = OtlpConfiguration(
            timeout: OtlpConfiguration.DefaultTimeoutInterval,
            headers: [("Authorization", "Bearer \(apiKey)")]
        )

        let exporter = OtlpHttpTraceExporter(
            endpoint: url,
            config: otlpConfig
        )

        // ----------------------------------------------------------------
        // 2. Wrap with a BatchSpanProcessor (buffer 512, schedule 1 s)
        // ----------------------------------------------------------------
        let batchConfig = BatchSpanProcessorConfiguration(
            maxExportBatchSize: 512,
            scheduledDelayMillis: 1_000,
            exportTimeoutMillis: 30_000,
            maxQueueSize: 2_048
        )
        let processor = BatchSpanProcessor(
            spanExporter: exporter,
            meterProvider: nil,
            configuration: batchConfig
        )

        // ----------------------------------------------------------------
        // 3. Create a TracerProvider with the service.name resource
        // ----------------------------------------------------------------
        let resource = Resource(attributes: [
            ResourceAttributes.serviceName: AttributeValue.string(service)
        ])

        tracerProvider = TracerProviderSdk(
            resource: resource,
            spanProcessors: [processor]
        )

        // Register as the global OTel provider so helper calls work.
        OpenTelemetry.registerTracerProvider(tracerProvider: tracerProvider)

        otelTracer = tracerProvider.get(
            instrumentationName: "tracep",
            instrumentationVersion: "1.0.0"
        )
    }

    // MARK: - Public API

    /// Begins a new root trace and returns an opaque trace-ID string.
    ///
    /// - Parameter name: Human-readable name for the root span.
    /// - Returns: Hex trace-ID that must be passed to subsequent calls.
    @discardableResult
    public func start(name: String) -> String {
        let rootSpan = otelTracer
            .spanBuilder(spanName: name)
            .setSpanKind(spanKind: .client)
            .startSpan()

        let traceIdHex = rootSpan.context.traceId.hexString

        let ctx = TraceContext(traceId: traceIdHex, rootSpan: rootSpan)

        traceLock.lock()
        activeTraces[traceIdHex] = ctx
        traceLock.unlock()

        return traceIdHex
    }

    /// Adds a span event to the span keyed by `cls + function` within `tid`.
    ///
    /// - Parameters:
    ///   - tid:      Trace-ID returned by `start()`.
    ///   - cls:      Class / module name (becomes `code.namespace` attribute).
    ///   - function: Function name (becomes `code.function` attribute).
    ///   - step:     Name of the span event.
    ///   - message:  Human-readable message attached to the event.
    ///   - parent:   Optional parent span key (`"\(cls).\(function)"` of parent span).
    ///   - level:    Severity string, default `"info"`.
    public func trace(
        tid: String,
        cls: String,
        function: String,
        step: String,
        message: String,
        parent: String? = nil,
        level: String = "info"
    ) {
        queue.async { [weak self] in
            guard let self else { return }

            self.traceLock.lock()
            let ctx = self.activeTraces[tid]
            self.traceLock.unlock()

            guard let ctx else {
                // Unknown tid — drop silently per spec.
                return
            }

            let spanKey = "\(cls).\(function)"

            // Resolve parent context if a parentKey was supplied.
            let parentSpan: Span?
            if let parentKey = parent {
                // If a specific parent span key is given, try to find it;
                // fall back to the root span.
                parentSpan = ctx.getOrCreateSpan(key: parentKey) {
                    // We should not create new spans for a parent hint that
                    // doesn't exist yet; return root as a safe fallback.
                    ctx.rootSpan
                }
            } else {
                parentSpan = nil
            }

            // Find or create the child span.
            let childSpan = ctx.getOrCreateSpan(key: spanKey) {
                var builder = self.otelTracer
                    .spanBuilder(spanName: "\(cls).\(function)")
                    .setSpanKind(spanKind: .internal)

                if let ps = parentSpan {
                    // Propagate the parent span's context.
                    builder = builder.setParent(ps)
                } else {
                    // Link to the root span.
                    builder = builder.setParent(ctx.rootSpan)
                }

                let span = builder.startSpan()
                span.setAttribute(key: "code.namespace", value: cls)
                span.setAttribute(key: "code.function",  value: function)
                return span
            }

            // Add a span event with message + level attributes.
            let eventAttributes: [String: OpenTelemetryApi.AttributeValue] = [
                "message": .string(message),
                "level":   .string(level),
            ]
            childSpan.addEvent(name: step, attributes: eventAttributes)
        }
    }

    /// Ends the trace identified by `tid`, flushing all spans to the exporter.
    ///
    /// - Parameters:
    ///   - tid:    Trace-ID returned by `start()`.
    ///   - status: `"ok"` (default) or `"error"`.
    public func end(tid: String, status: String) {
        queue.async { [weak self] in
            guard let self else { return }

            self.traceLock.lock()
            let ctx = self.activeTraces.removeValue(forKey: tid)
            self.traceLock.unlock()

            guard let ctx else { return }

            ctx.endAll(status: status)

            // Force-flush so events are exported before the caller moves on.
            _ = self.tracerProvider.forceFlush(timeout: 5)
        }
    }

    /// Shuts down the tracer provider, flushing any remaining spans.
    /// Call once when the process is about to exit.
    public func close() {
        queue.sync {
            // End any still-active traces as "ok".
            traceLock.lock()
            let remaining = activeTraces
            activeTraces.removeAll()
            traceLock.unlock()

            for (_, ctx) in remaining {
                ctx.endAll(status: "ok")
            }

            _ = tracerProvider.forceFlush(timeout: 10)
            tracerProvider.shutdown()
        }
    }
}

// ---------------------------------------------------------------------------
// MARK: - Retry-aware export (wraps exporter with 3× exponential back-off)
// ---------------------------------------------------------------------------

/// A `SpanExporter` decorator that retries up to `maxAttempts` times with
/// exponential back-off, then drops the batch silently on final failure.
final class RetrySpanExporter: SpanExporter {

    private let wrapped: SpanExporter
    private let maxAttempts: Int
    private let baseDelaySeconds: Double

    init(wrapped: SpanExporter, maxAttempts: Int = 3, baseDelaySeconds: Double = 1.0) {
        self.wrapped = wrapped
        self.maxAttempts = maxAttempts
        self.baseDelaySeconds = baseDelaySeconds
    }

    func export(spans: [SpanData], explicitTimeout: TimeInterval?) -> SpanExporterResultCode {
        var attempt = 0
        while attempt < maxAttempts {
            let result = wrapped.export(spans: spans, explicitTimeout: explicitTimeout)
            if result == .success {
                return .success
            }
            attempt += 1
            if attempt < maxAttempts {
                let delay = baseDelaySeconds * pow(2.0, Double(attempt - 1))
                Thread.sleep(forTimeInterval: delay)
            }
        }
        // Drop silently after maxAttempts.
        return .failure
    }

    func flush() -> SpanExporterResultCode { wrapped.flush() }
    func shutdown() { wrapped.shutdown() }
}
